### PR TITLE
First cut at TermEnum API for discovering terms in the index.

### PR DESCRIPTION
**Pull request by markharwood**
Patterns can be provided with a variety of flags similar to those used in term, prefix, wildcard and regexp queries.
Matching terms can be returned in alphabetical order or by popularity,
A timeout can limit the amount of time spent looking for matches.

Expected to be useful in auto-complete or regex debugging use cases.

### CODE REVIEW

File: docs/reference/search/term-enum.asciidoc [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+[source,console]
+--------------------------------------------------
+POST stackoverflow/_terms/list
+{
+    "field" : "tags",
+    "pattern" : "kiba"
+}
+--------------------------------------------------
+// TEST[setup:stackoverflow]
```
Comment (mayya-sharipova) [inline docs/reference/search/term-enum.asciidoc:15]: An Interesting API, is it supposed to be externally facing as well? Or only be used internally for suggestions etc?

File: docs/reference/search/term-enum.asciidoc [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+[[search-term-enum-api-request-body]]
+==== {api-request-body-title}
+
+[[term-enum-field-param]]
+`field`::
+(Mandatory, string)
+Which field to match
+
+[[term-enum-field-param]]
```
Comment (mayya-sharipova) [inline docs/reference/search/term-enum.asciidoc:73]: should this be called `pattern`?

File: docs/reference/search/term-enum.asciidoc (hunk 1 above)
Comment (markharwood) [reply docs/reference/search/term-enum.asciidoc:15]: > is it supposed to be externally facing as well? 

I'd like it to be public facing - it seems perverse to hide the terms that prefix, wildcard and regexp queries use under the covers. Why shouldn't we be transparent about what's in the index? 
Would stop questions [like this](https://discuss.elastic.co/t/help-elasticsearch-regexp-query/253445) which I've seen many times.

**Comment by markharwood** [comment]
Closing while we rethink the API

**Pull request closed without merging**